system/lib/pthread/proxying.c

```diff
@@ -0,0 +1,326 @@
+    int new_capacity = tasks->capacity * 2;
+    task* new_tasks = malloc(sizeof(task) * new_capacity);
+    if (new_tasks == NULL) {
+      return 0;
+    }
+    // Copy the tasks such that the head of the queue is at the beginning of the
+    // buffer. There are two cases to handle: either the queue wraps around the
+    // end of the old buffer or it does not.
+    int queued_tasks;
+    if (tasks->head <= tasks->tail) {
+      // No wrap. Copy the tasks in one chunk.
+      queued_tasks = tasks->tail - tasks->head;
+      memcpy(
+        new_tasks, &tasks->tasks[tasks->head], sizeof(task) * queued_tasks);
+    } else {
+      // Wrap. Copy `first_queued` tasks up to the end of the old buffer and
+      // `last_queued` tasks at the beginning of the old buffer.
+      int first_queued = tasks->capacity - tasks->head;
+      int last_queued = tasks->tail;
+      queued_tasks = first_queued + last_queued;
+      memcpy(
+        new_tasks, &tasks->tasks[tasks->head], sizeof(task) * first_queued);
+      memcpy(
+        new_tasks + first_queued, tasks->tasks, sizeof(task) * last_queued);
+    }
+    free(tasks->tasks);
+    tasks->tasks = new_tasks;
+    tasks->capacity = new_capacity;
+    tasks->head = 0;
+    tasks->tail = queued_tasks;
+  }
+  // Enqueue the task.
+  tasks->tasks[tasks->tail] = t;
+  tasks->tail = (tasks->tail + 1) % tasks->capacity;
+  return 1;
+}
+
+static task task_queue_dequeue(task_queue* tasks) {
+  task t = tasks->tasks[tasks->head];
+  tasks->head = (tasks->head + 1) % tasks->capacity;
+  return t;
+}
+
+struct em_proxying_queue {
+  // Protects all accesses to all task_queues.
+  pthread_mutex_t mutex;
+  // `size` task queues stored in an array of size `capacity`.
+  task_queue* task_queues;
+  int size;
+  int capacity;
+};
+
+static em_proxying_queue system_proxying_queue = {
+  PTHREAD_MUTEX_INITIALIZER, NULL, 0, 0};
+
+em_proxying_queue* emscripten_proxy_get_system_queue(void) {
+  return &system_proxying_queue;
+}
+
+em_proxying_queue* em_proxying_queue_create(void) {
+  em_proxying_queue* q = malloc(sizeof(em_proxying_queue));
+  if (q == NULL) {
+    return NULL;
+  }
+  pthread_mutex_init(&q->mutex, NULL);
+  q->task_queues = NULL;
+  q->size = 0;
+  q->capacity = 0;
+  return q;
+}
+
+void em_proxying_queue_destroy(em_proxying_queue* q) {
+  if (q == NULL) {
+    return;
+  }
+  assert(q != &system_proxying_queue && "cannot destroy system proxying queue");
+  // No need to acquire the lock; no one should be racing with the destruction
+  // of the queue.
+  pthread_mutex_destroy(&q->mutex);
+  for (int i = 0; i < q->size; i++) {
+    free(q->task_queues[i].tasks);
+  }
+  free(q->task_queues);
+  free(q);
+}
+
+// Not thread safe.
+static task_queue* get_tasks_for_thread(em_proxying_queue* q,
+                                        pthread_t thread) {
+  assert(q != NULL);
+  task_queue* tasks = NULL;
+  for (int i = 0; i < q->size; i++) {
+    if (pthread_equal(q->task_queues[i].thread, thread)) {
```

Comment:
Why not use TLS here to avoid the linear search?   I.e. can't `task_queues` be accessed by `pthread_get_specific`?

Reply:
That would work for looking up the queue for the current thread, but is there a way to make that work for looking up the queue for another thread? I guess we could have both TLS and the array, but that seems more complicated for unclear gain. I expect the linear search to very fast anyway, since I don't imagine there would ever be more than a small handful of target threads.